src/Dibi/Drivers/PdoDriver.php

```diff
@@ -89,31 +89,44 @@ public function disconnect(): void
 	 */
 	public function query(string $sql): ?Dibi\ResultDriver
 	{
-		$res = $this->connection->query($sql);
-		if ($res) {
-			$this->affectedRows = $res->rowCount();
-			return $res->columnCount() ? $this->createResultDriver($res) : null;
+		try {
+			$res = $this->connection->query($sql);
+
+			if ($res) {
+				$this->affectedRows = $res->rowCount();
+				return $res->columnCount() ? $this->createResultDriver($res) : null;
+			}
+
+		} catch (\PDOException $pdoException) {
+			$this->affectedRows = null;
+			throw $this->createException($pdoException->errorInfo, $sql, $pdoException);
 		}
 
 		$this->affectedRows = null;
+		throw $this->createException($this->connection->errorInfo(), $sql, null);
```

Comment:
I do not like this handling of error cases. Is is done twice, any ideas?

Reply:
Is your intention to make it work regardless of ATTR_ERRMODE, ie. with ERRMODE_WARNING and ERRMODE_EXCEPTION together?

Maybe you can use finally:

```php
try {
    $res = $this->connection->query($sql);
    ...

} catch (\PDOException $e) {
} finally {
	throw $this->createException(isset($e) ?  $e->errorInfo : $this->connection->errorInfo(), $sql);
}
```

Reply:
Finally will be executed also on successful case, I have extracted this in main body of the function.